Index: org.eclipse.imp.metatooling/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.metatooling/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.metatooling/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-SymbolicName: org.eclipse.imp.metatooling;singleton:=true
 Bundle-Name: IMP Language Support Meta-Tooling (Incubation)
 Bundle-Vendor: rfuhrer@watson.ibm.com
-Bundle-Version: 0.1.92
+Bundle-Version: 0.1.92.qualifier
 Bundle-Activator: org.eclipse.imp.WizardPlugin
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
Index: org.eclipse.imp.lpg/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.lpg/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.lpg/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-ManifestVersion: 2
 Bundle-Name: LPG IDE & Meta-tooling (Incubation)
 Bundle-SymbolicName: org.eclipse.imp.lpg
-Bundle-Version: 0.1.84
+Bundle-Version: 0.1.84.qualifier
 Bundle-Activator: org.eclipse.imp.lpg.Activator
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/language/LanguageRegistry.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/language/LanguageRegistry.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/language/LanguageRegistry.java	(working copy)
@@ -21,6 +21,8 @@
 import java.util.List;
 import java.util.Map;
 
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.ResourcesPlugin;
 import org.eclipse.core.runtime.IConfigurationElement;
 import org.eclipse.core.runtime.IExtensionPoint;
 import org.eclipse.core.runtime.IPath;
@@ -134,7 +136,15 @@
 	public static Language findLanguage(IEditorInput editorInput, IDocumentProvider docProvider) {
 		if (!sIsFullyInitialized)
 			initializeRegistryAsNeeded();
-		IPath path= EditorInputUtils.getPath(editorInput);
+		
+		// LK: get full path
+		IFile file = EditorInputUtils.getFile(editorInput);
+		IPath path;
+		if (file != null) {
+			path = file.getFullPath();
+		} else {
+			path = EditorInputUtils.getPath(editorInput);
+		}
 
 		return findLanguage(path, docProvider.getDocument(editorInput));
 	}
@@ -164,7 +174,9 @@
 		            LanguageValidator validator = lang.getValidator();
 
 		            if (validator != null && docContents != null) {
-		                if (validator.validate(docContents)) {
+		            	// LK: needed for MetaFileLanguageValidator
+		            	IFile file = ResourcesPlugin.getWorkspace().getRoot().getFile(path);
+		            	if (validator.validate(file) ||validator.validate(docContents)) {
 		                    return lang;
 		                }
 		            } else {
@@ -383,7 +395,10 @@
         for (String ext : extensions) {
 	        IFileEditorMapping mapping= findMappingFor(ext, mappings);
 
-	        if (mapping == null || mapping.getDefaultEditor().getId().equals(sUniversalEditor.getId())) {
+	        if (mapping == null
+	        		|| (!(mapping instanceof IMPFileEditorMapping) // LK: mapping may already be a IMPFileEditorMapping 
+	        			 && (mapping.getDefaultEditor() == null
+	        		         || mapping.getDefaultEditor().getId().equals(sUniversalEditor.getId())))) {
 	            // Replace the file editor mapping even if it already pointed to the universal editor,
 	            // since the persisted association turns into a FileEditorMapping when re-read, thus
 	            // losing the icon (which FileEditorMapping gets from the IEditorDescriptor).
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/services/base/TreeModelBuilderBase.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/services/base/TreeModelBuilderBase.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/services/base/TreeModelBuilderBase.java	(working copy)
@@ -23,7 +23,8 @@
 
     private Stack<ModelTreeNode> fItemStack= new Stack<ModelTreeNode>();
 
-    public final ModelTreeNode buildTree(Object rootASTNode) {
+    // LK: public interface shouldn't be final
+    public ModelTreeNode buildTree(Object rootASTNode) {
         fItemStack.push(fModelRoot= createTopItem(new ModelTreeNode(rootASTNode)));
         try {
             visitTree(rootASTNode);
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/editor/ServiceControllerManager.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/editor/ServiceControllerManager.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/editor/ServiceControllerManager.java	(working copy)
@@ -48,6 +48,11 @@
         if (fLanguageServiceManager.getHyperLinkDetector() != null)
             fHyperLinkController= new SourceHyperlinkController(fLanguageServiceManager.getHyperLinkDetector(), fTextEditor);
     }
+    
+    public ServiceControllerManager() { // LK: TODO: REMOVE ME
+    	fTextEditor = null;
+    
+    }
 
     public void initialize() {
         IRegionSelectionService regionSelector= (IRegionSelectionService) fTextEditor.getAdapter(IRegionSelectionService.class);
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/editor/StructuredSourceViewer.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/editor/StructuredSourceViewer.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/editor/StructuredSourceViewer.java	(working copy)
@@ -7,7 +7,6 @@
 *
 * Contributors:
 *    Robert Fuhrer (rfuhrer@watson.ibm.com) - initial API and implementation
-
 *******************************************************************************/
 
 package org.eclipse.imp.editor;
@@ -16,6 +15,7 @@
 
 import org.eclipse.imp.editor.UniversalEditor.StructuredSourceViewerConfiguration;
 import org.eclipse.imp.parser.IParseController;
+import org.eclipse.imp.runtime.RuntimePlugin;
 import org.eclipse.imp.services.IAutoEditStrategy;
 import org.eclipse.imp.services.ILanguageSyntaxProperties;
 import org.eclipse.imp.services.base.DefaultAutoIndentStrategy;
@@ -35,6 +35,7 @@
 import org.eclipse.swt.graphics.Color;
 import org.eclipse.swt.graphics.Point;
 import org.eclipse.swt.widgets.Composite;
+// LK: backported from revision 22255
 
 public class StructuredSourceViewer extends ProjectionViewer {
     /**
@@ -63,7 +64,6 @@
      */
     public static final int MARK_OCCURRENCES= 55;
 
-    // mmk 4/8/08
     /**
      * Text operation code for indenting the currently selected text.
      */
@@ -77,16 +77,8 @@
 
     private IAutoEditStrategy fAutoEditStrategy;
 
-    // SMS 29 May 2007
     private IParseController fParseController;
     
-    /**
-     * Is this source viewer configured?
-     *
-     * @since 3.0
-     */
-    private boolean fIsConfigured;
-
     public StructuredSourceViewer(Composite parent, IVerticalRuler verticalRuler, IOverviewRuler overviewRuler, boolean showAnnotationsOverview, int styles) {
         super(parent, verticalRuler, overviewRuler, showAnnotationsOverview, styles);
     }
@@ -139,8 +131,7 @@
         }
         super.doOperation(operation);
     }
-    
-    // SMS 29 May 2007 (see doToggleComment())
+
     public void setParseController(IParseController parseController) {
     	fParseController = parseController;
     }
@@ -152,13 +143,14 @@
     private void doToggleComment() {
         ILanguageSyntaxProperties syntaxProps= fParseController.getSyntaxProperties();
 
-        if (syntaxProps == null)
+        // LK: single line comment prefix can be null
+        if (syntaxProps == null || syntaxProps.getSingleLineCommentPrefix() == null)
             return;
 
         IDocument doc= this.getDocument();
         DocumentRewriteSession rewriteSession= null;
         Point p= this.getSelectedRange();
-        final String lineCommentStart= syntaxProps.getSingleLineCommentPrefix();
+        final String lineCommentPrefix= syntaxProps.getSingleLineCommentPrefix();
 
     	if (doc instanceof IDocumentExtension4) {
     	    IDocumentExtension4 extension= (IDocumentExtension4) doc;
@@ -172,22 +164,35 @@
             final int startLine= doc.getLineOfOffset(selStart);
             int endLine= doc.getLineOfOffset(selEnd);
 
-            if (selLen > 0 && doc.getChar(selEnd - 1) == '\n')
+            if (selLen > 0 && lookingAtLineEnd(doc, selEnd))
                 endLine--;
+
+            boolean linesAllHaveCommentPrefix= linesHaveCommentPrefix(doc, lineCommentPrefix, startLine, endLine);
+        	boolean useCommonLeadingSpace= true; // take from a preference?
+			int leadingSpaceToUse= useCommonLeadingSpace ? calculateLeadingSpace(doc, startLine, endLine) : 0;
+
             for(int line= startLine; line <= endLine; line++) {
-                int lineStartOffset= doc.getLineOffset(line);
-                int offset= lineStartOffset;
+                int lineStart= doc.getLineOffset(line);
+                int lineEnd= lineStart + doc.getLineLength(line) - 1;
 
-                while (Character.isWhitespace(doc.getChar(offset)) && offset < doc.getLength())
-                    offset++;
-                if (doc.get(offset, lineCommentStart.length()).equals(lineCommentStart)) {
-                    int len= lineCommentStart.length();
-
-                    while (Character.isWhitespace(doc.getChar(offset + len)))
-                        len++;
-                    doc.replace(offset, len, "");
-                } else
-                    doc.replace(offset, 0, lineCommentStart + " ");
+                if (linesAllHaveCommentPrefix) {
+                	// remove the comment prefix from each line, wherever it occurs in the line
+                	int offset= lineStart;
+                    while (Character.isWhitespace(doc.getChar(offset)) && offset < lineEnd) {
+                        offset++;
+                    }
+                    // LK: remove additional space after comment prefix
+                    int offsetEnd = offset + lineCommentPrefix.length();
+					if (offsetEnd < lineEnd && doc.getChar(offsetEnd) == ' ') {
+						doc.replace(offset, lineCommentPrefix.length() + 1, "");
+					} else {
+						doc.replace(offset, lineCommentPrefix.length(), "");
+					}
+                } else {
+                	// add the comment prefix to each line, after however many spaces leadingSpaceToAdd indicates
+                	int offset= lineStart + leadingSpaceToUse;
+                	doc.replace(offset, 0, lineCommentPrefix + " ");
+                }
             }
         } catch (BadLocationException e) {
             e.printStackTrace();
@@ -200,8 +205,54 @@
         }
     }
 
-    // mmk 4/8/08
-    private void doIndentLines() {
+    private int calculateLeadingSpace(IDocument doc, int startLine, int endLine) {
+    	try {
+        	int result= Integer.MAX_VALUE;
+        	for(int line= startLine; line <= endLine; line++) {
+        		int lineStart= doc.getLineOffset(line);
+        		int lineEnd= lineStart + doc.getLineLength(line) - 1;
+        		int offset= lineStart;
+        		while (Character.isWhitespace(doc.getChar(offset)) && offset < lineEnd) {
+        			offset++;
+        		}
+        		int leadingSpaces= offset - lineStart;
+				result= Math.min(result, leadingSpaces);
+        	}
+    		return result;
+    	} catch (BadLocationException e) {
+    		return 0;
+    	}
+	}
+
+	/**
+     * @return true, if the given inclusive range of lines all start with the single-line comment prefix,
+     * even if they have different amounts of leading whitespace
+     */
+    private boolean linesHaveCommentPrefix(IDocument doc, String lineCommentPrefix, int startLine, int endLine) {
+    	try {
+    		int docLen= doc.getLength();
+
+    		for(int line= startLine; line <= endLine; line++) {
+                int lineStart= doc.getLineOffset(line);
+                int lineEnd= lineStart + doc.getLineLength(line) - 1;
+                int offset= lineStart;
+
+                while (Character.isWhitespace(doc.getChar(offset)) && offset < lineEnd) {
+                    offset++;
+                }
+                if (docLen - offset > lineCommentPrefix.length() && doc.get(offset, lineCommentPrefix.length()).equals(lineCommentPrefix)) {
+                	// this line starts with the single-line comment prefix
+                } else {
+                	return false;
+                }
+            }
+    	} catch (BadLocationException e) {
+    		return false;
+    	}
+		return true;
+	}
+
+	private void doIndentLines() {
         IDocument doc= this.getDocument();
         DocumentRewriteSession rewriteSession= null;
         Point p= this.getSelectedRange();
@@ -218,18 +269,16 @@
             final int startLine= doc.getLineOfOffset(selStart);
             int endLine= doc.getLineOfOffset(selEnd);
 
-            if (selLen > 0 && doc.getChar(selEnd - 1) == '\n')
+            if (selLen > 0 && lookingAtLineEnd(doc, selEnd))
                 endLine--;
             for(int line= startLine; line <= endLine; line++) {
                 int lineStartOffset= doc.getLineOffset(line);
-                int offset= lineStartOffset;
 
-                // Replace the indent with desired indent
-                // make use of the language-specific AutoEditStrategy
-                // This strategy assumes keystroke input, so infrastructure requires a key event/DocumentCommand to work.
+                // Replace the existing indentation with the desired indentation.
+                // Use the language-specific AutoEditStrategy, which requires a DocumentCommand.
                 DocumentCommand cmd= new DocumentCommand() { };
                 cmd.offset= lineStartOffset;
-                cmd.length= 1;
+                cmd.length= 0;
                 cmd.text= Character.toString('\t');
                 cmd.doit= true;
                 cmd.shiftsCaret= false;
@@ -237,7 +286,7 @@
                 doc.replace(cmd.offset, cmd.length, cmd.text);
             }
         } catch (BadLocationException e) {
-            e.printStackTrace();
+            RuntimePlugin.getInstance().logException("Indent Selection command failed", e);
         } finally {
             if (doc instanceof IDocumentExtension4) {
                 IDocumentExtension4 extension= (IDocumentExtension4) doc;
@@ -247,6 +296,21 @@
         }
     }
 
+    private boolean lookingAtLineEnd(IDocument doc, int pos) {
+        String[] legalLineTerms= doc.getLegalLineDelimiters();
+        try {
+            for(String lineTerm: legalLineTerms) {
+                int len= lineTerm.length();
+                if (pos > len && doc.get(pos - len, len).equals(lineTerm)) {
+                    return true;
+                }
+            }
+        } catch (BadLocationException e) {
+            RuntimePlugin.getInstance().logException("Error examining document for line termination", e);
+        }
+        return false;
+    }
+
     /*
      * @see ISourceViewer#configure(SourceViewerConfiguration)
      */
@@ -296,12 +360,15 @@
                 		fAutoEditStrategy = new DefaultAutoIndentStrategy();
                 }
             }
+            
+            fQuickAssistAssistant = sSVConfiguration.getQuickAssistAssistant(this);
+            if (fQuickAssistAssistant != null)
+            	fQuickAssistAssistant.install(this);
         }
         //	if (fPreferenceStore != null) {
         //	    fPreferenceStore.addPropertyChangeListener(this);
         //	    initializeViewerColors();
         //	}
-        fIsConfigured= true;
     }
 
     /*
@@ -332,6 +399,5 @@
         //	if (fPreferenceStore != null)
         //	    fPreferenceStore.removePropertyChangeListener(this);
         super.unconfigure();
-        fIsConfigured= false;
     }
 }
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/editor/ParserScheduler.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/editor/ParserScheduler.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/editor/ParserScheduler.java	(working copy)
@@ -29,6 +29,7 @@
 import org.eclipse.jface.text.IDocument;
 import org.eclipse.ui.IEditorInput;
 import org.eclipse.ui.IEditorPart;
+import org.eclipse.ui.texteditor.AbstractTextEditor;
 import org.eclipse.ui.texteditor.IDocumentProvider;
 
 /**
@@ -37,7 +38,7 @@
  */
 // TODO Perhaps this should be driven off of the "IReconcilingStrategy" mechanism?
 public class ParserScheduler extends Job {
-    private final IParseController fParseController;
+    private IParseController fParseController;
 
     private final IEditorPart fEditorPart;
 
@@ -64,7 +65,8 @@
     }
 
     public IStatus run(IProgressMonitor monitor) {
-        if (fParseController == null || fDocumentProvider == null) {
+    	 // LK: test if document provider is null to find out if editor was closed
+        if (fParseController == null || ((AbstractTextEditor) fEditorPart).getDocumentProvider() == null) {
             /* Editor was closed, or no parse controller */
             return Status.OK_STATUS;
         }
@@ -82,10 +84,12 @@
             // Just make sure the document contents gets parsed once (and only once).
             fMsgHandler.clearMessages();
             fParseController.parse(document.get(), monitor);
-            if (!monitor.isCanceled())
+            if (!monitor.isCanceled()) {
                 notifyModelListeners(monitor);
+            }
         } catch (Exception e) {
-            ErrorHandler.reportError("Error running parser for language " + fParseController.getLanguage().getName() + " and input " + editorInput.getName() + ":", e);
+        	// LK: null check
+            ErrorHandler.reportError("Error running parser for language " + (fParseController == null ? "null" : fParseController.getLanguage().getName()) + " and input " + editorInput.getName() + ":", e);
             // RMF 8/2/2006 - Notify the AST listeners even on an exception - the compiler front end
             // may have failed at some phase, but there may be enough info to drive IDE services.
             notifyModelListeners(monitor);
@@ -146,4 +150,8 @@
             RuntimePlugin.getInstance().writeInfoMsg("No AST; bypassing listener notification.");
         }
     }
+
+	public void dispose() { // LK
+		fParseController = null;
+	}
 }
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/editor/internal/FormattingController.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/editor/internal/FormattingController.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/editor/internal/FormattingController.java	(working copy)
@@ -33,7 +33,8 @@
     }
 
     public void formatterStarts(String initialIndentation) {
-        fFormattingStrategy.formatterStarts(initialIndentation);
+    	if (fFormattingStrategy != null)
+    		fFormattingStrategy.formatterStarts(initialIndentation);
     }
 
     public String format(String content, boolean isLineStart, String indentation, int[] positions) {
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/editor/internal/PresentationController.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/editor/internal/PresentationController.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/editor/internal/PresentationController.java	(working copy)
@@ -13,6 +13,7 @@
 
 import java.io.PrintStream;
 import java.util.ArrayList;
+import java.util.EmptyStackException;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Stack;
@@ -37,6 +38,7 @@
 import org.eclipse.swt.SWT;
 import org.eclipse.swt.custom.StyleRange;
 import org.eclipse.swt.widgets.Display;
+import org.eclipse.ui.PlatformUI;
 
 /**
  * A class that does the real work of repairing the text presentation for an associated ISourceViewer.
@@ -109,22 +111,38 @@
         IRegion bigRegion= fColorer.calculateDamageExtent(region, fParseCtlr);
 
         if (bigRegion != null) {
+        	// LK: (no changes, just a comment)
+        	// Implicit synchronization here risks a deadlock in the colorer implementation,
+        	// which might need main thread access to create new SWT Color Objects
             fWorkItems.push(bigRegion);
         }
     }
 
     public void update(IParseController controller, IProgressMonitor monitor) {
-        if (!monitor.isCanceled()) {
+        // LK: only run in the UI thread so color objects are created there
+    	if (PlatformUI.getWorkbench().getActiveWorkbenchWindow() == null)
+        	return;
+    	if (!monitor.isCanceled()) {
 //            if (fWorkItems.size() == 0) {
 //                ConsoleUtil.findConsoleStream(PresentationController.CONSOLE_NAME).println("PresentationController.update() called, but no damage in the work queue?");
 //            }
-            synchronized (fWorkItems) {
-                for(int n= fWorkItems.size() - 1; !monitor.isCanceled() && n >= 0; n--) {
-                    Region damage= (Region) fWorkItems.get(n);
-                    changeTextPresentationForRegion(controller, monitor, damage);
-                }
-                if (!monitor.isCanceled())
-                    fWorkItems.removeAllElements();
+        	// LK: locking using fWorkItems as syncroot can result in deadlocks
+            synchronized (this) {
+            	try {
+	                while (!monitor.isCanceled() && !fWorkItems.isEmpty()) {
+	                    Region damage = (Region) fWorkItems.pop();
+	                    synchronized (fWorkItems) { // LK: clean up work items queue
+	                    	while (!fWorkItems.isEmpty()
+	                    			&& (fWorkItems.peek().getOffset() >= damage.getOffset() && fWorkItems.peek().getLength() <= damage.getLength()))
+	                    		fWorkItems.pop();
+	                    }
+	                    changeTextPresentationForRegion(controller, monitor, damage);
+	                }
+	                // if (!monitor.isCanceled())
+	                //    fWorkItems.removeAllElements();
+            	} catch (EmptyStackException e) {
+            		// Ignore
+            	}
             }
         }
     }
@@ -157,7 +175,7 @@
             int offset= locator.getStartOffset(token);
             int end= locator.getEndOffset(token);
 
-            if (offset <= prevEnd && end >= prevOffset) {
+            if (offset <= prevEnd /*&& end >= prevOffset*/) { // LK: Avoid illegal styleRange
                 continue;
             }
             changeTokenPresentation(parseController, presentation, token, locator);
@@ -176,7 +194,7 @@
         
        // SMS 21 Jun 2007:  negative (possibly 0) length style ranges seem to cause problems;
         // but if you have one it should lead to an IllegalArgumentException in changeTextPresentation(..)
-        if (styleRange.length <= 0 || styleRange.start + styleRange.length > this.fSourceViewer.getDocument().getLength()) {
+        if (styleRange.length <= 0 ||  this.fSourceViewer.getDocument() == null || styleRange.start + styleRange.length > this.fSourceViewer.getDocument().getLength()) {
         } else {
             presentation.addStyleRange(styleRange);
         }
@@ -190,6 +208,7 @@
             	// SMS 16 Sep 2008
             	int charCount;
             	if (fSourceViewer != null) {
+            		if (fSourceViewer.getDocument() == null) return; // LK: handle closed documents better
             		charCount = fSourceViewer.getDocument().getLength();
             	} else {
             		charCount = 0;
@@ -201,7 +220,7 @@
             	int lastLength = presentation.getLastStyleRange().length;
             	int end = lastStart + lastLength;
         		TextPresentation newPresentation = null;
-            	if (end >= charCount) {
+            	if (end > charCount) { // LK: > not >=
             		newPresentation = new TextPresentation();
             		Iterator presIt = presentation.getAllStyleRangeIterator();
             		while (presIt.hasNext()) {
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/editor/UniversalEditor.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/editor/UniversalEditor.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/editor/UniversalEditor.java	(working copy)
@@ -11,10 +11,13 @@
 
 package org.eclipse.imp.editor;
 
+import static org.eclipse.ui.texteditor.AbstractDecoratedTextEditorPreferenceConstants.EDITOR_TAB_WIDTH;
+
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
@@ -31,7 +34,10 @@
 import org.eclipse.core.runtime.CoreException;
 import org.eclipse.core.runtime.IPath;
 import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
 import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.core.runtime.Status;
 import org.eclipse.debug.ui.actions.IToggleBreakpointsTarget;
 import org.eclipse.debug.ui.actions.ToggleBreakpointAction;
 import org.eclipse.help.IContextProvider;
@@ -42,6 +48,7 @@
 import org.eclipse.imp.editor.internal.AnnotationCreator;
 import org.eclipse.imp.editor.internal.EditorErrorTickUpdater;
 import org.eclipse.imp.editor.internal.FoldingController;
+import org.eclipse.imp.editor.internal.PresentationController;
 import org.eclipse.imp.editor.internal.ProblemMarkerManager;
 import org.eclipse.imp.editor.internal.ToggleBreakpointsAdapter;
 import org.eclipse.imp.help.IMPHelp;
@@ -163,6 +170,7 @@
 import org.eclipse.ui.editors.text.TextSourceViewerConfiguration;
 import org.eclipse.ui.handlers.IHandlerActivation;
 import org.eclipse.ui.handlers.IHandlerService;
+import org.eclipse.ui.progress.UIJob;
 import org.eclipse.ui.texteditor.AbstractTextEditor;
 import org.eclipse.ui.texteditor.ContentAssistAction;
 import org.eclipse.ui.texteditor.IEditorStatusLine;
@@ -240,6 +248,9 @@
     private static final String BUNDLE_FOR_CONSTRUCTED_KEYS= MESSAGE_BUNDLE;//$NON-NLS-1$
 
     private static final String IMP_EDITOR_CONTEXT= RuntimePlugin.IMP_RUNTIME + ".imp_editor_context";
+    
+    // LK
+    private final Set<IDocumentListener> onSaveListeners = new HashSet<IDocumentListener>();
 
     static ResourceBundle fgBundleForConstructedKeys= ResourceBundle.getBundle(BUNDLE_FOR_CONSTRUCTED_KEYS);
 
@@ -257,6 +268,41 @@
         setInsertMode(SMART_INSERT);
         fProblemMarkerManager= new ProblemMarkerManager();
     }
+    
+    public ServiceControllerManager getServiceControllerManager() { // LK
+        return fServiceControllerManager;
+    }
+    
+    public LanguageServiceManager getLanguageServiceManager() { // LK
+        return fLanguageServiceManager;
+    }
+    
+    public IPreferenceStore getThePreferenceStore() { // LK
+    	return super.getPreferenceStore();
+    }
+    
+    public void addOnSaveListener(IDocumentListener listener) { // LK
+    	onSaveListeners.add(listener);
+    }
+    
+    public void removeOnSaveListener(IDocumentListener listener) { // LK
+    	onSaveListeners.remove(listener);
+    }
+    
+    public void updateColoring(Region region) { // LK
+    	final PresentationController presentation = fServiceControllerManager.getPresentationController();
+		presentation.damage(region);
+		// Must run in main thread in order not to keep further updates waiting
+		UIJob job = new UIJob("Update presentation") {
+			@Override
+			public IStatus runInUIThread(IProgressMonitor monitor) {
+				presentation.update(fLanguageServiceManager.getParseController(), monitor);
+				return Status.OK_STATUS;
+			}
+		};
+		job.setSystem(true);
+		job.schedule();
+    }
 
     public Object getAdapter(Class required) {
         if (IContentOutlinePage.class.equals(required)) {
@@ -696,13 +742,14 @@
         IEditorInput editorInput= getEditorInput();
         IFile file= EditorInputUtils.getFile(editorInput);
         IPath filePath= EditorInputUtils.getPath(editorInput);
+    	// LK: initialize parse controller even if project cannot be determined
+        ISourceProject srcProject = null;
         try {
-            ISourceProject srcProject= (file != null) ? ModelFactory.open(file.getProject()) : null;
-
-            fLanguageServiceManager.getParseController().initialize(filePath, srcProject, fAnnotationCreator);
+           srcProject = ModelFactory.open(file.getProject());
         } catch (ModelException e) {
             ErrorHandler.reportError("Error initializing parser for input " + editorInput.getName() + ":", e);
         }
+        fLanguageServiceManager.getParseController().initialize(filePath, srcProject, fAnnotationCreator);
     }
 
     private void findLanguageSpecificPreferences() {
@@ -981,7 +1028,7 @@
         }
     }
     
-    private static final int REPARSE_SCHEDULE_DELAY= 100;
+    private static final int REPARSE_SCHEDULE_DELAY= 75;
 
     private void setTitleImageFromLanguageIcon() {
         // Only set the editor's title bar icon if the language has a label provider
@@ -1099,6 +1146,10 @@
           fActionBars = null;
         }
         
+        if (fParserScheduler != null) { // LK
+        	fParserScheduler.dispose();
+        }
+        
         super.dispose();
     }
 
@@ -1470,7 +1521,13 @@
         public int getTabWidth(ISourceViewer sourceViewer) {
             boolean langSpecificSetting= fLangSpecificPrefs != null && fLangSpecificPrefs.isDefined(PreferenceConstants.P_TAB_WIDTH);
 
-            return langSpecificSetting ? fLangSpecificPrefs.getIntPreference(PreferenceConstants.P_TAB_WIDTH) : PreferenceCache.tabWidth;
+            if (langSpecificSetting) {
+            	return fLangSpecificPrefs.getIntPreference(PreferenceConstants.P_TAB_WIDTH);
+            } else {
+            	// LK: instead of using PreferenceCache.tabWidth, use one of the language-generic settings of Eclipse
+        		IPreferenceStore preferences = getThePreferenceStore();
+        		return preferences.getInt(EDITOR_TAB_WIDTH);
+            }
         }
 
         public IPresentationReconciler getPresentationReconciler(ISourceViewer sourceViewer) {
@@ -1491,6 +1548,7 @@
             ContentAssistant ca= new ContentAssistant();
             ca.setContentAssistProcessor(fServiceControllerManager.getCompletionProcessor(), IDocument.DEFAULT_CONTENT_TYPE);
             ca.setInformationControlCreator(getInformationControlCreator(sourceViewer));
+            ca.enableColoredLabels(true); // LK
             return ca;
         }
 
@@ -1771,7 +1829,8 @@
                 if (fServiceControllerManager.getPresentationController() != null) {
 //                  System.out.println("Scheduling repair for damage to region " + damage.getOffset() + ":" + damage.getLength() + " in doc of length " + fDocument.getLength());
                     fServiceControllerManager.getPresentationController().damage(damage);
-                    if (hyperlinkRestore) {
+                    // LK: just always color for post-marker color correctness
+                    if (hyperlinkRestore || true) {
 //                      System.out.println("** Forcing repair for hyperlink damage to region " + damage.getOffset() + ":" + damage.getLength() + " in doc of length " + fDocument.getLength());
                         fServiceControllerManager.getPresentationController().update(fLanguageServiceManager.getParseController(), fProgressMonitor);
                     }
@@ -1845,6 +1904,7 @@
     }
 
     public IParseController getParseController() {
+    	if (fLanguageServiceManager == null) return null; // LK
         return fLanguageServiceManager.getParseController();
     }
     
@@ -1861,7 +1921,23 @@
 		// SMS 25 Apr 2007:  Removing parser annotations here
 		// may not hurt but also doesn't seem to be necessary
 		//removeParserAnnotations();
+		DocumentEvent event = null;
+		try {
+			event = new DocumentEvent(getServiceControllerManager().getSourceViewer().getDocument(), 0, 0, null);
+			for (IDocumentListener listener : onSaveListeners)
+				listener.documentAboutToBeChanged(event);
+		} catch (RuntimeException e) {
+			e.printStackTrace();
+		}
+		
 		super.doSave(progressMonitor);
+		
+		try {
+			for (IDocumentListener listener : onSaveListeners)
+				listener.documentChanged(event);
+		} catch (RuntimeException e) {
+			e.printStackTrace();
+		}
 	}
 
     public void removeParserAnnotations() {
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/editor/SourceProposal.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/editor/SourceProposal.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/editor/SourceProposal.java	(working copy)
@@ -134,7 +134,9 @@
 
     public void apply(IDocument document) {
         try {
-            document.replace(fRange.getOffset(), fRange.getLength(), fNewText.substring(fPrefix.length()));
+        	// LK: use accessors
+            Region range = getRange();
+			document.replace(range.getOffset(), range.getLength(), getNewText().substring(getPrefix().length()));
         } catch (BadLocationException e) {
             e.printStackTrace();
         }
@@ -155,6 +157,18 @@
     public Image getImage() {
         return null;
     }
+    
+    public String getNewText() { // LK
+		return fNewText;
+	}
+    
+    public Region getRange() { // LK
+		return fRange;
+	}
+    
+    public String getPrefix() { // LK
+		return fPrefix;
+	}
 
     public IContextInformation getContextInformation() {
         return null;
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/utils/AnnotationUtils.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/utils/AnnotationUtils.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/utils/AnnotationUtils.java	(working copy)
@@ -21,6 +21,7 @@
 import org.eclipse.jface.text.Position;
 import org.eclipse.jface.text.source.Annotation;
 import org.eclipse.jface.text.source.IAnnotationModel;
+import org.eclipse.jface.text.source.ILineDiffInfo;
 import org.eclipse.jface.text.source.ISourceViewer;
 import org.eclipse.jface.text.source.projection.AnnotationBag;
 
@@ -30,6 +31,7 @@
 
     public static String formatAnnotationList(List<Annotation> annotations) {
         if (annotations != null) {
+            removeLineDiffInfo(annotations); // LK
             if (annotations.size() == 1) {
                 // optimization
                 Annotation annotation= (Annotation) annotations.get(0);
@@ -52,6 +54,15 @@
         return null;
     }
 
+	private static void removeLineDiffInfo(List<Annotation> annotations) { // LK
+		if (annotations == null) return;
+		for (Iterator<Annotation> iter = annotations.iterator(); iter.hasNext(); ) {
+    		Annotation annotation = iter.next();
+    		if (annotation.getText() == null || annotation instanceof ILineDiffInfo)
+    			iter.remove();
+    	}
+	}
+	
     public static IAnnotationModel getAnnotationModel(ISourceViewer viewer) {
         // if (viewer instanceof ISourceViewerExtension2) {
         // ISourceViewerExtension2 extension= (ISourceViewerExtension2) viewer;
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/preferences/PreferenceCache.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/preferences/PreferenceCache.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/preferences/PreferenceCache.java	(working copy)
@@ -17,7 +17,7 @@
 public class PreferenceCache {
     public static boolean emitMessages= false;
 
-    public static int tabWidth= 8;
+    public static int tabWidth= 4; // LK: sane default tab width
 
     public static Font sourceFont;
 
Index: org.eclipse.imp.runtime/src/org/eclipse/imp/preferences/PreferenceInitializer.java
===================================================================
--- org.eclipse.imp.runtime/src/org/eclipse/imp/preferences/PreferenceInitializer.java	(revision 22255)
+++ org.eclipse.imp.runtime/src/org/eclipse/imp/preferences/PreferenceInitializer.java	(working copy)
@@ -42,7 +42,7 @@
         if (fontData != null && fontData.length > 0)
             PreferenceConverter.setDefault(store, PreferenceConstants.P_SOURCE_FONT, fontData);
 
-        store.setDefault(PreferenceConstants.P_TAB_WIDTH, 8);
+        store.setDefault(PreferenceConstants.P_TAB_WIDTH, 4); // LK: sane default tab width
         store.setDefault(PreferenceConstants.P_DUMP_TOKENS, false);
         store.setDefault(PreferenceConstants.EDITOR_MATCHING_BRACKETS, true);
 
Index: org.eclipse.imp.runtime/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.runtime/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.runtime/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-ManifestVersion: 2
 Bundle-Name: IMP Runtime (Incubation)
 Bundle-SymbolicName: org.eclipse.imp.runtime; singleton:=true
-Bundle-Version: 0.1.101
+Bundle-Version: 0.1.101.qualifier
 Bundle-Activator: org.eclipse.imp.runtime.RuntimePlugin
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
@@ -30,6 +30,7 @@
  org.eclipse.imp.builder,
  org.eclipse.imp.core,
  org.eclipse.imp.editor,
+ org.eclipse.imp.editor.internal,
  org.eclipse.imp.indexing,
  org.eclipse.imp.language,
  org.eclipse.imp.model,
Index: org.eclipse.imp.java.hosted/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.java.hosted/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.java.hosted/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-ManifestVersion: 2
 Bundle-Name: IMP Support for Java-Hosted Languages (Incubation)
 Bundle-SymbolicName: org.eclipse.imp.java.hosted
-Bundle-Version: 0.1.2
+Bundle-Version: 0.1.2.qualifier
 Bundle-Activator: org.eclipse.imp.java.hosted.Activator
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
Index: org.eclipse.imp.presentation/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.presentation/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.presentation/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-ManifestVersion: 2
 Bundle-Name: IMP Presentation Specifications (Incubation)
 Bundle-SymbolicName: org.eclipse.imp.presentation;singleton:=true
-Bundle-Version: 0.1.7
+Bundle-Version: 0.1.7.qualifier
 Bundle-Activator: org.eclipse.imp.presentation.PSPActivator
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
Index: org.eclipse.imp.lpg.ide/plugin.xml
===================================================================
--- org.eclipse.imp.lpg.ide/plugin.xml	(revision 22255)
+++ org.eclipse.imp.lpg.ide/plugin.xml	(working copy)
@@ -244,15 +244,6 @@
       </entityNameLocator>
    </extension>
    <extension
-         point="org.eclipse.ui.preferencePages">
-      <page
-            name="LPG"
-            class="org.eclipse.imp.lpg.preferences.LPGPreferencePage"
-            id="org.eclipse.imp.lpg.preferences.LPG">
-      </page>
-   </extension>
-
-   <extension
          id="docBuilder"
          name="LPG Grammar Documentation Builder"
          point="org.eclipse.core.resources.builders">
@@ -265,9 +256,18 @@
    <extension
          point="org.eclipse.help.toc">
       <toc
-            category="IMP"
             file="doc/toc.xml"
-            primary="true">
+            primary="true"
+            category="IMP">
       </toc>
    </extension>
+   <extension
+         point="org.eclipse.ui.preferencePages">
+      <page
+            name="LPG"
+            class="org.eclipse.imp.lpg.preferences.LPGPreferencePage"
+            id="org.eclipse.imp.lpg.preferences.LPG">
+      </page>
+   </extension>
+
 </plugin>
Index: org.eclipse.imp.lpg.ide/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.lpg.ide/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.lpg.ide/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-ManifestVersion: 2
 Bundle-Name: LPG IDE Runtime (Incubation)
 Bundle-SymbolicName: org.eclipse.imp.lpg.ide;singleton:=true
-Bundle-Version: 0.1.87
+Bundle-Version: 0.1.87.qualifier
 Bundle-Activator: org.eclipse.imp.lpg.LPGRuntimePlugin
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
@@ -12,7 +12,6 @@
  org.eclipse.core.runtime,
  org.eclipse.core.resources,
  org.eclipse.core.variables,
- org.eclipse.help.ui;bundle-version="3.4.0",
  org.eclipse.imp.runtime,
  org.eclipse.imp.smapifier,
  org.eclipse.imp.xform,
Index: org.eclipse.imp.prefspecs/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.prefspecs/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.prefspecs/META-INF/MANIFEST.MF	(working copy)
@@ -4,7 +4,7 @@
 Bundle-Activator: org.eclipse.imp.prefspecs.PrefspecsPlugin
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
-Bundle-Version: 0.1.89
+Bundle-Version: 0.1.89.qualifier
 Require-Bundle: lpg.runtime,
  org.eclipse.core.runtime,
  org.eclipse.core.resources,
Index: org.eclipse.imp.xform/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.xform/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.xform/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-ManifestVersion: 2
 Bundle-Name: IMP Transformation Toolkit (Incubation)
 Bundle-SymbolicName: org.eclipse.imp.xform; singleton:=true
-Bundle-Version: 0.1.91
+Bundle-Version: 0.1.91.qualifier
 Bundle-Activator: org.eclipse.imp.xform.XformPlugin
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
Index: org.eclipse.imp.preferences/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.preferences/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.preferences/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-ManifestVersion: 2
 Bundle-Name: IMP Preferences Meta-tooling (Incubation)
 Bundle-SymbolicName: org.eclipse.imp.preferences;singleton:=true
-Bundle-Version: 0.1.86
+Bundle-Version: 0.1.86.qualifier
 Bundle-Activator: org.eclipse.imp.preferences.PreferencesPlugin
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
Index: org.eclipse.imp.smapi/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.smapi/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.smapi/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-ManifestVersion: 2
 Bundle-Name: SMAP support (Incubation)
 Bundle-SymbolicName: org.eclipse.imp.smapi
-Bundle-Version: 0.1.89.2
+Bundle-Version: 0.1.89.qualifier
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
 Export-Package: org.eclipse.imp.smapi
Index: org.eclipse.imp.smapifier/META-INF/MANIFEST.MF
===================================================================
--- org.eclipse.imp.smapifier/META-INF/MANIFEST.MF	(revision 22255)
+++ org.eclipse.imp.smapifier/META-INF/MANIFEST.MF	(working copy)
@@ -2,7 +2,7 @@
 Bundle-ManifestVersion: 2
 Bundle-Name: SMAP Builder (Incubation)
 Bundle-SymbolicName: org.eclipse.imp.smapifier;singleton:=true
-Bundle-Version: 0.1.89.2
+Bundle-Version: 0.1.89.qualifier
 Bundle-Activator: org.eclipse.imp.smapifier.SmapiePlugin
 Bundle-ActivationPolicy: lazy
 Eclipse-LazyStart: true
